clang/lib/CConv/ConstraintResolver.cpp

```diff
@@ -679,6 +696,18 @@ CVarSet ConstraintResolver::getBaseVarPVConstraint(DeclRefExpr *Decl) {
   return Ret;
 }
 
+// Construct a PVConstraint for an expression that can safely be used when
+// rewriting the expression later on. This is done by making the constraint WILD
+// if the expression is inside a macro.
+PVConstraint *ConstraintResolver::getRewritablePVConstraint(Expr *E) {
+  PVConstraint *P = new PVConstraint(E->getType(), nullptr,
+                                     E->getStmtClassName(), Info, *Context,
+                                     nullptr);
+  CVarSet Tmp = {P};
+  Info.constrainWildIfMacro(Tmp, E->getExprLoc());
```

Comment:
Rather than make a `CV` that you know you will not use (since you've constrained it to be WILD), would it make sense to fail (return `nullptr`) and thus force the caller to not create it in the first place? You could also do the "is rewritable" check here, and fail for that reason, too, maybe?

Reply:
I don't think so. The WILD constrained CV is returned up the call stack in `getExprConstraintVars`, so it causes other expressions to be WILD. 

```c
#define foo(x) ((int**)x)
void test() {
  int **c = foo(0);
}
```

With the current implementation, `c` is made WILD. If this method returned `nullptr`(or perhaps an empty CVars set), `c` would be made `_Ptr<_Ptr<int>>` which fails to compile because we cannot rewrite the cast inside the macro. 

What might work (haven't tested this out yet) is returning `getWildPVConstraint()` here instead. Do you think this approach would be better?

Reply:
What you are describing may not be what I had in mind. I am saying that `ConstraintResolver::getRewritablePVConstraint` return `nullptr` and then the caller reacts to that. It may be that the caller needs to create a wild pointer CV, or it may be able to avoid creating a CV altogether.

I may be off base, so I'm OK to leave it as is.

Reply:
I think that the caller will always need to constrain the result to wild, at least for how I intend this function to be used.